File: app/src/main/java/at/shockbytes/dante/dagger/AppModule.java
package at.shockbytes.dante.dagger;

import android.app.Application;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.google.gson.ExclusionStrategy;
import com.google.gson.FieldAttributes;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import javax.inject.Named;
import javax.inject.Singleton;

import at.shockbytes.dante.backup.google.GoogleSignInManager;
import at.shockbytes.dante.network.BookDownloader;
import at.shockbytes.dante.network.google.gson.BookBackupSerializer;
import at.shockbytes.dante.network.google.gson.GoogleBooksSuggestionResponseDeserializer;
import at.shockbytes.dante.util.AppParams;
import at.shockbytes.dante.util.DanteRealmMigration;
import at.shockbytes.dante.backup.BackupManager;
import at.shockbytes.dante.backup.GoogleDriveBackupManager;
import at.shockbytes.dante.util.books.BookManager;
import at.shockbytes.dante.util.books.BookSuggestion;
import at.shockbytes.dante.util.books.RealmBookManager;
import at.shockbytes.dante.util.tracking.KeenTracker;
import at.shockbytes.dante.util.tracking.Tracker;
import dagger.Module;
import dagger.Provides;
import io.realm.Realm;
import io.realm.RealmConfiguration;
import io.realm.RealmObject;

/**
 * @author Martin Macheiner
 *         Date: 13.02.2017.
 */

@Module
public class AppModule {

    private Application app;

    public AppModule(Application app) {
        this.app = app;
    }

    @Provides
    @Singleton
    public SharedPreferences provideSharedPreferences() {
        return PreferenceManager.getDefaultSharedPreferences(app.getApplicationContext());
    }

    @Provides
    @Singleton
    public BookManager provideBookManager(BookDownloader bookDownloader, Realm realm) {
        return new RealmBookManager(bookDownloader, realm);
    }

    @Provides
    @Singleton
    public Gson provideGson() {
        return new GsonBuilder()
                .registerTypeAdapter(BookSuggestion.class, new GoogleBooksSuggestionResponseDeserializer())
                .create();
    }

    @Provides
    @Singleton
    @Named("backup_gson")
    public Gson provideBackupGson() {
        try {
            return new GsonBuilder()
                    .setExclusionStrategies(new ExclusionStrategy() {
                        @Override
                        public boolean shouldSkipField(FieldAttributes f) {
                            return f.getDeclaringClass().equals(RealmObject.class);
                        }

                        @Override
                        public boolean shouldSkipClass(Class<?> clazz) {
                            return false;
                        }
                    })
                    .registerTypeAdapter(Class.forName("io.realm.BookRealmProxy"), new BookBackupSerializer())
                    .create();
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        }
        return new Gson(); // This should never be the case
    }

    @Provides
    @Singleton
    public Tracker provideTracker() {
        return new KeenTracker(app.getApplicationContext());
    }

    @Provides
    @Singleton
    public Realm provideRealm() {
        RealmConfiguration config = new RealmConfiguration.Builder()
                .schemaVersion(AppParams.REALM_SCHEMA_VERSION)
                .migration(new DanteRealmMigration())
                .build();
        return Realm.getInstance(config);
    }

    @Provides
    @Singleton
    public BackupManager provideBackupManager(SharedPreferences preferences,
                                              GoogleSignInManager signInManager,
                                              @Named("backup_gson") Gson gson) {
        return new GoogleDriveBackupManager(preferences, signInManager, gson);
    }

    @Provides
    @Singleton
    GoogleSignInManager provideGoogleSignInManager(SharedPreferences prefs) {
        return new GoogleSignInManager(prefs);
    }

}

File: app/src/main/java/at/shockbytes/dante/dagger/AppModule.kt
package at.shockbytes.dante.dagger

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import at.shockbytes.dante.BuildConfig
import at.shockbytes.dante.backup.BackupRepository
import at.shockbytes.dante.backup.DefaultBackupRepository
import at.shockbytes.dante.backup.provider.BackupProvider
import at.shockbytes.dante.backup.provider.google.GoogleDriveBackupProvider
import at.shockbytes.dante.backup.provider.shockbytes.ShockbytesHerokuServerBackupProvider
import at.shockbytes.dante.backup.provider.shockbytes.api.ShockbytesHerokuApi
import at.shockbytes.dante.backup.provider.shockbytes.storage.InactiveShockbytesBackupStorage
import at.shockbytes.dante.backup.provider.shockbytes.storage.SharedPreferencesInactiveShockbytesBackupStorage
import at.shockbytes.dante.book.BookSuggestion
import at.shockbytes.dante.book.realm.RealmInstanceProvider
import at.shockbytes.dante.network.google.gson.GoogleBooksSuggestionResponseDeserializer
import at.shockbytes.dante.signin.GoogleSignInManager
import at.shockbytes.dante.signin.SignInManager
import at.shockbytes.dante.ui.image.GlideImageLoader
import at.shockbytes.dante.ui.image.ImageLoader
import at.shockbytes.dante.ui.image.ImagePicker
import at.shockbytes.dante.ui.image.RxLegacyImagePicker
import at.shockbytes.dante.util.DanteRealmMigration
import at.shockbytes.dante.util.settings.DanteSettings
import at.shockbytes.dante.flagging.FeatureFlagging
import at.shockbytes.dante.flagging.FirebaseFeatureFlagging
import at.shockbytes.dante.flagging.SharedPreferencesFeatureFlagging
import at.shockbytes.dante.util.scheduler.AppSchedulerFacade
import at.shockbytes.dante.util.scheduler.SchedulerFacade
import com.google.firebase.remoteconfig.FirebaseRemoteConfig
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import dagger.Module
import dagger.Provides
import dagger.Reusable
import io.realm.RealmConfiguration
import javax.inject.Named
import javax.inject.Singleton

/**
 * Author:  Martin Macheiner
 * Date:    13.02.2017
 */
@Module
class AppModule(private val app: Application) {

    @Provides
    @Singleton
    fun provideApplication(): Application {
        return app
    }

    @Provides
    @Singleton
    fun provideSharedPreferences(): SharedPreferences {
        return PreferenceManager.getDefaultSharedPreferences(app.applicationContext)
    }

    @Provides
    @Singleton
    fun provideDanteSettings(sharedPreferences: SharedPreferences): DanteSettings {
        return DanteSettings(app.applicationContext, sharedPreferences)
    }

    @Provides
    @Singleton
    @Named("gsonDownload")
    fun provideGson(): Gson {
        return GsonBuilder()
                .registerTypeAdapter(BookSuggestion::class.java, GoogleBooksSuggestionResponseDeserializer())
                .create()
    }

    @Provides
    @Singleton
    fun provideRealmInstanceProvider(): RealmInstanceProvider {
        return RealmInstanceProvider(RealmConfiguration.Builder()
                .schemaVersion(DanteRealmMigration.migrationVersion)
                .migration(DanteRealmMigration())
                .build())
    }

    @Provides
    @Reusable
    fun provideInactiveShockbytesBackupStorage(
        preferences: SharedPreferences
    ): InactiveShockbytesBackupStorage {
        return SharedPreferencesInactiveShockbytesBackupStorage(preferences)
    }

    @Provides
    @Reusable
    fun provideBackupRepository(
        backupProvider: Array<BackupProvider>,
        preferences: SharedPreferences
    ): BackupRepository {
        return DefaultBackupRepository(backupProvider.toList(), preferences)
    }

    @Provides
    @Reusable
    fun provideBackupProvider(
        schedulerFacade: SchedulerFacade,
        signInManager: SignInManager,
        shockbytesHerokuApi: ShockbytesHerokuApi,
        inactiveShockbytesBackupStorage: InactiveShockbytesBackupStorage
    ): Array<BackupProvider> {
        return arrayOf(
            GoogleDriveBackupProvider(
                signInManager as GoogleSignInManager,
                schedulerFacade,
                Gson()
            ),
            ShockbytesHerokuServerBackupProvider(
                signInManager,
                shockbytesHerokuApi,
                inactiveShockbytesBackupStorage
            )
        )
    }

    @Provides
    @Singleton
    fun provideGoogleSignInManager(prefs: SharedPreferences): SignInManager {
        return GoogleSignInManager(prefs, app.applicationContext)
    }

    @Provides
    @Singleton
    fun provideFeatureFlagging(remoteConfig: FirebaseRemoteConfig): FeatureFlagging {
        return if (BuildConfig.DEBUG) {
            val prefs = app.getSharedPreferences("feature_flagging", Context.MODE_PRIVATE)
            SharedPreferencesFeatureFlagging(prefs)
        } else {
            FirebaseFeatureFlagging(remoteConfig)
        }
    }

    @Provides
    @Singleton
    fun provideImageLoader(): ImageLoader {
        return GlideImageLoader
    }

    @Provides
    @Singleton
    fun provideSchedulerFacade(): SchedulerFacade {
        return AppSchedulerFacade()
    }

    @Provides
    @Singleton
    fun provideImagePicker(): ImagePicker {
        return RxLegacyImagePicker()
    }
}
